opentelemetry-sdk/src/metrics/internal/sum.rs

```diff
@@ -19,89 +20,122 @@ use super::{aggregate::is_under_cardinality_limit, AtomicTracker, Number};
 pub(crate) static STREAM_OVERFLOW_ATTRIBUTES: Lazy<Vec<KeyValue>> =
     Lazy::new(|| vec![KeyValue::new("otel.metric.overflow", "true")]);
 
+/// Abstracts the update operation for a measurement.
+trait Operation {
```

Comment:
Another way to implement a customizable behavior for update operation would be to use a function pointer. We could add a new field to the `ValueMap` struct and set the desired behavior for update during the struct creation and use the function pointer to execute the update logic.

```rust
struct ValueMap<T: Number<T>> {
   ...
   update: fn(&T::AtomicTracker, T),
}

impl<T: Number<T>> ValueMap<T> {
    fn new(is_sum_precomputed: bool) -> Self {
        ValueMap {
            values: RwLock::new(HashMap::new()),
            has_no_value_attribute_value: AtomicBool::new(false),
            no_attribute_value: T::new_atomic_tracker(),
            count: AtomicUsize::new(0),
            update: if is_sum_precomputed { Self::store } else { Self::add },
        }
    }

    fn store(tracker: &T::AtomicTracker, measurement: T) {
        tracker.store(measurement);
    }

    fn add(tracker: &T::AtomicTracker, measurement: T) {
        tracker.add(measurement);
    }
}

// Use (self.update)(&value_to_update, measurement); to update the atomic tracker value in the measure method.
```

@stormshield-fabs What are your thoughts on using generics vs function pointers? Any interesting pros or cons?

Reply:
I think this works too and is much more readable! 
It might introduce an additional call indirection (since the compiler can inline the generic version) but it's probably not super impactful.

I can update this PR to use function pointers if the other reviewers agree

Reply:
Generic's advantage of allowing inline is worth keeping the generic approach.. (Its hard to measure the impact with our benchmarks i think.. this is most likely within noise level of that!)

Let us go with the generic now, and revisit as we work through rest of Metrics SDK issues.